Adds goldmaster test

theNerd247 opened the pull request:
I'm adding a gold master test as part of our unit test to ensure that future refactors maintain legacy API behavior and features. Once breaking API changes are introduced we'll need to either: A) update the goldmaster test data to use the new API or B) replace the goldmaster test with proper unit-tests

theNerd247 committed:
add gold-master test for legacy fnr-text api

A simple gold-master test has been added. The tests forego searching and
replacing via regex due to the  complexities involved with generating
arbitrary regular expressions.

theNerd247 committed:
set min string length in unit test

theNerd247 committed:
makes nonSubStringOf use asciiString

kintsoogi committed:
Vscode wants to fix spacing

kintsoogi committed:
One more formatting change

kintsoogi committed:
Added code review comments

kintsoogi reviewed (commented):
I added some code comments, idk if you want me to request changes, comment or approve, so I just commented

theNerd247 committed:
add description for goldmaster testing process

theNerd247 committed:
adds link to fast-check introduction

theNerd247 committed:
adds comment about single letter import aliases

theNerd247 committed:
refactor goldmaster test code to SearchAndReplaceArbitrary

I've refactored this bit into it's own module and switched to using
node's fs/promises API to read/write to the json file. I did this to
ensure that the file we were using for the goldmaster data was consistent
on both reading and writing and to make the usage of the arbitrary code
in that module a bit more apparent.

theNerd247 committed:
removes commented out code and adds a few descriptions

kintsoogi committed:
Commented regarding T as import

Responded to the comment about importing types at T from index. I found the goldmaster explanation and the medium article on property-based testing helpful. Also, the extra comments in SearchAndReplaceArb were very helpful

theNerd247 committed:
adds response to Chris

theNerd247 committed:
moves index.d.ts file to a more meaningful file name

theNerd247 committed:
re-adds index.d.ts

theNerd247 committed:
removes discussion on import naming conventions

kintsoogi reviewed (approved):
Alright it all looks good to me! The commenting process was helpful!
